### PR TITLE
Backend Jest tests.

### PR DESCRIPTION
**Pull request by diversifiedmechanic**
I'm really sorry about this being so long.  I should have split it up into smaller pieces and added a test at a time!

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,4 @@
 /node_modules
 /.env
 /public/bundle.js
+coverage
```
Comment (ZacharyAb) [inline .gitignore]: GJ adding your coverage file here, you do seem to be missing a new line at the end of your page which is normally requested in my repo, not sure if your's is the same in that regard.

File: __tests__/dbAndSeeding.test.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    const insertedItems = await addManyItems(generatedItems)
+    expect(insertedItems.length).toBe(generatedItems.length);
+  });
+
+  it('should return all the items by a seller except the current item', async () => {
+    const allExceptCurrent = await getAllSellerItemsExceptCurrentItem(1);
+    const arrayOfItemIds = allExceptCurrent.map((item) => (item.itemId));
+    expect(arrayOfItemIds).not.toContain(1);
+  });
+});
```
Comment (ZacharyAb) [inline __tests__/dbAndSeeding.test.js]: lgtm, great tests here. I had to reread what you're testing for on line 47 a couple of times, I don't believe you need to test to make sure your array is empty on each loop if you're redeclaring it as an empty array.

File: database/db.js [hunk 3]
```diff
@@ -0,0 +1,8 @@
+require('dotenv').config();
+// mongoose set up and connection
+const mongoose = require('mongoose');
+
+mongoose.connect(process.env.DATABASE_URL, { useUnifiedTopology: true, useNewUrlParser: true });
+
+
+module.exports.db = mongoose.connection;
```
Comment (ZacharyAb) [inline database/db.js]: lgtm

File: database/index.js [hunk 4]
```diff
@@ -48,5 +20,5 @@ module.exports.getAllSellerItemsExceptCurrentItem = (idOfItem) => (
 );
 
 module.exports.addManyItems = (arrayOfItems) => (
-  Schemas.Item.insertMany(arrayOfItems)
+  Item.insertMany(arrayOfItems)
 );
```
Comment (ZacharyAb) [inline database/index.js]: Wow so many great clean changes here! I like the direction you're moving in.

File: database/index.test.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+  it('should return all items except the one given', async () => {
+    const returnedItems = await getAllSellerItemsExceptCurrentItem(1);
+    const idArray = returnedItems.map((item) => item.itemId);
+    expect(idArray).not.toContain(1);
+
+    const returnedItems2 = await getAllSellerItemsExceptCurrentItem(4);
+    const idArray2 = returnedItems2.map((item) => item.itemId);
+    expect(idArray2).not.toContain(4);
+  });
+});
```
Comment (ZacharyAb) [inline database/index.test.js]: Your tests are really clear in what you're trying to accomplish here, lgtm!

File: database/schemas.test.js [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    const invalidItem = new Item({ itemId: 2, itemLocation: 'UnitedStates' });
+    const savedInvalidItem = await invalidItem.save();
+    expect(savedInvalidItem.itemId).toBe(invalidItem.itemId);
+    expect(savedInvalidItem.itemLocation).toBeUndefined();
+  });
+
+  // additional tests could include:
+  // making all fields required
+  // value types are matching, String = a string
+});
```
Comment (ZacharyAb) [inline database/schemas.test.js]: Thorough, I like the notes on additional testing for future and that you tested for not inserting an invalid field. Your code is clean and clearly well thought out here.

File: dummyData/dataSeeder.js [hunk 7]
```diff
@@ -50,17 +49,10 @@ const databaseSeeder = (numOfSellers = 10, numOfItemsPerSeller = 10) => {
...
-      process.exit();
-    });
+  return arrayOfFakeData;
 };
 
-databaseSeeder();
+
+module.exports = {
+  generateData,
+};
```
Comment (ZacharyAb) [inline dummyData/dataSeeder.js]: Nicely cleaned up.

File: dummyData/seeder.js [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+      process.exit();
+    })
+    .catch((error) => {
+      // console logs the error then excess the node process
+      console.error('Seeding the data had the following error:', error);
+      process.exit();
+    });
+};
+
+populateDatabase();
```
Comment (ZacharyAb) [inline dummyData/seeder.js]: Nice separation of concerns with the actual seeding function, lgtm

File: dummyData/testData.js [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+    sellerJoinDate: '2019-10-21T23:17:39.897Z',
+    sellerPicture: 'https://www.placecage.com/300/310',
+    itemId: 10,
+    itemName: 'Rustic Wooden Keyboard',
+    itemPrice: 712,
+    itemPicture: 'https://i.picsum.photos/id/519/300/310.jpg',
+    itemShippingPrice: 43,
+    __v: 0,
+  },
+];
```
Comment (ZacharyAb) [inline dummyData/testData.js]: lgtm

File: jest-mongodb-config.js [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+    instance: {
+      dbName: 'jest',
+    },
+    binary: {
+      version: '4.0.3',
+      skipMD5: true,
+    },
+    autoStart: false,
+  },
+};
```
Comment (ZacharyAb) [inline jest-mongodb-config.js]: Don't know enough about setting up Mongo with Jest to have a great comment here, but on initial look seems fine.

File: package.json [hunk 11]
```diff
@@ -28,21 +28,28 @@
...
     "eslint-plugin-import": "^2.20.0",
     "eslint-plugin-jsx-a11y": "^6.2.3",
     "eslint-plugin-react": "^7.18.0",
     "faker": "^4.1.0",
+    "jest": "^25.1.0",
     "nodemon": "^2.0.2",
+    "supertest": "^4.0.2",
     "webpack-cli": "^3.3.10"
   }
 }
```
Comment (ZacharyAb) [inline package.json]: lgtm

File: server/app.js [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+  db.getAllSellerItemsExceptCurrentItem(req.params.id)
+    .then((allItems) => {
+      res.status(200).json(allItems);
+    })
+    .catch(() => {
+      res.status(404).json({ error: 'invalid input!' });
+    });
+});
+
+module.exports = app;
```
Comment (ZacharyAb) [inline server/app.js]: Good use of the catch block here, interesting to see you not pass the error and instead send it with the 404.

File: server/app.test.js [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+        expect(data.length).toBe(9);
+      })
+  });
+
+  it('should return with a 500 error if something went wrong', async () => {
+    await endpoint
+      .get('/Foobar')
+      .expect(404)
+  })
+});
```
Comment (ZacharyAb) [inline server/app.test.js]: Solid server test, unclear why it should be returning a 500 error when something went wrong but you're expecting a 404.

File: server/index.js [hunk 14]
```diff
@@ -1,24 +1,6 @@
...
-    .then((allItems) => {
-      res.status(200).json(allItems);
-    })
-    .catch((error) => {
-      res.status(500).json(error);
-    });
-});
-
-
 app.listen(port, console.log(`Listening on port ${port}`));
```
Comment (ZacharyAb) [inline server/index.js]: Nice separation of concerns.

File: .gitignore (hunk 1 above)
Comment (diversifiedmechanic) [reply .gitignore]: That's a good point.  I think I'm ignoring it in my ESLint and I shouldn't be.

File: __tests__/dbAndSeeding.test.js (hunk 2 above)
Comment (diversifiedmechanic) [reply __tests__/dbAndSeeding.test.js]: I believe that I'm testing to make sure that all of the items have unique names.  I fill the array, checking for duplicates until I hit a new seller.  At that point I reset the array to empty then start checking for duplicates again.

File: jest-mongodb-config.js (hunk 10 above)
Comment (diversifiedmechanic) [reply jest-mongodb-config.js]: It's for my 'in memory mongo db' I have some extra files because of it.  I should go back and see which ones I actually need because I'm sure some of them are redundant.  Thanks for reminding me!

File: server/app.js (hunk 12 above)
Comment (diversifiedmechanic) [reply server/app.js]: yeah.... not sure this was the best call.  I'm probably going to re-write how my promises are working so I can send different errors depending on where the problem went wrong.

File: server/app.test.js (hunk 13 above)
Comment (diversifiedmechanic) [reply server/app.test.js]: haha oh man.  Thanks for the catch!  This is the exact moment I decided to switch up how my server is responding to errors.  This is going to be changed.
